### ISSUE TITLE
Batched cone program canonicalization

### ISSUE THREAD

**Issue by bamos** [cvxgrp/cvxpylayers#53]
Serializing the canonicalization takes a non-negligible amount of time for larger batches (~500ms for a batch size of 128) and is relatively easy to batch up for cone programs. I've sped this up to ~100ms by hacking in a manual batched canonicalization by pulling the sparse affine maps for the CP and applying them to the batch of parameters by replacing the call to `apply_parameters`  [here](https://github.com/cvxgrp/cvxpylayers/blob/ffce695/cvxpylayers/torch/cvxpylayer.py#L234) with the snippet below (which at the moment doesn't handle some edge cases as carefully as `apply_parameters`).

Where would the best place for code like this to live? Does it make sense add something like `apply_parameters_batch` to the cvxpy compiler that defaults to a serialized version but can optionally be over-ridden?

```Python
for i in range(batch_info.max_size):
    params_np_i = [
        p if sz == 0 else p[i]
        for p, sz in zip(params_np, batch_info.sizes)]
    d = dict(zip(self.param_ids, params_np_i))

    def param_value(idx):
        return d[idx]

    p = canonInterface.get_parameter_vector(
        self.compiler.total_param_size,
        self.compiler.param_id_to_col,
        self.compiler.param_id_to_size,
        param_value,
        zero_offset=False)
    params.append(p)

params = np.vstack(params).T
cs = self.compiler.c.dot(params).T
cs = cs[:,:-1]

params_sp = sp.csr_matrix(params)
Fs = self.compiler.A.dot(params_sp).T
As = -Fs[:,:-self.n]
As = [Ai.reshape(self.m, self.n).T for Ai in As]
bs = Fs[:,-self.n:].toarray()
```

\cc @SteveDiamond @priyald17

**Comment by akshayka** [comment]
Hey Brandon!

I suppose code like this could eventually live in CVXPY. Your instinct is right, in that CVXPY Layers shouldn't rely on the internals of the `compiler` object, since those are subject to change.

We're still ironing out basic performance issues in the compiler object in CVXPY. For example, the changes [in this PR](https://github.com/cvxgrp/cvxpy/pull/1010) dramatically speed up the creation of the data matrix, and (sometimes) dramatically speed up methods involved in the backward pass. In a test problem of my own, canonicalization was reduced from 550 ms to 10 ms. Depending on the nature/size of your problem, maybe you'll see similar speed-ups.

We could eventually add batch support to CVXPY. We'd have to be careful with the details, though.

For example, something like

```
Fs = self.compiler.A.dot(params_sp).T
```

won't always work, because `Fs` is dense here (sparse @ dense == dense). For large problems, you'll get OOM'd pretty quickly. The PR I linked to creates a reduced representation of `A`, by eliminating rows that are zero. As a result, we can do a dense matvec like you've done here, since the output is the data of the solver's CSC matrix.

Anyway, TL;DR, batch support would be nice. I don't have the bandwidth to add it & test it right now. And (maybe) the changes I've mentioned will get you far enough. Curious to hear what @SteveDiamond thinks.

**Comment by bamos** [comment]
I'm also now hitting a bottleneck in serializing the canonicalization derivatives and am adding in similar batching here.

> the changes [in this PR](https://github.com/cvxgrp/cvxpy/pull/1010) dramatically speed up the creation of the data matrix

I just pulled this PR and the performance is still about the same on my smaller problems

**Comment by SteveDiamond** [comment]
Maybe we could expand apply_parameters to take in a matrix of parameters (one column for each item in the batch)? @bamos is this what you mean by batching?

**Comment by bamos** [comment]
> Maybe we could expand apply_parameters to take in a matrix of parameters (one column for each item in the batch)? @bamos is this what you mean by batching?

Yeah, this works well too

**Comment by SteveDiamond** [comment]
The new cvxpylayers architecture applies parameters itself (rather than calling cvxpy) with matrix multiplication.
